### DISCUSSION TITLE
Versioned configs schemas (& general vision for how to scale projen)

### DISCUSSION THREAD

**Discussion by janeklb** [projen/projen#2983] [General]
First off.. this is a neat tool :)
When I first saw it, I was a bit skeptical of adding another layer of abstraction to my configs; however, after using it a little bit I am starting to appreciate certain scenarios where `projen` can really shine.

Before I start to lean on it too heavily, I would to ask about how `projen` plans on supporting configs across multiple versions of the tools it has constructs for.

For example, given that projen currently has support for generating / managing TypeScript 5 projects, consider:
- user creates TypeScript project with projen 🎉
- TypeScript 6 is released, and tsconfig v6 is not compatible with tsconfig v5 🚧
- does `projen` support both config schemas? 🤔 
- ... and now scale out across all other the versions of all the tools projen has constructs for..  🤯 

At this point I'm thinking you'd need some sort of versioned projen construct registry (eg. https://registry.terraform.io/)... would love to hear if this "problem" has been considered

**Comment by mrgrain** [comment]
> When I first saw it, I was a bit skeptical of adding another layer of abstraction to my configs; however, after using it a little bit I am starting to appreciate certain scenarios where projen can really shine.

Thank you! We appreciate the praise, but maybe even more so your constructive questions! ❤️ 

> Before I start to lean on it too heavily, I would to ask about how projen plans on supporting configs across multiple versions of the tools it has constructs for.
> [...]
> At this point I'm thinking you'd need some sort of versioned projen construct registry (eg. https://registry.terraform.io/)... would love to hear if this "problem" has been considered

I am aware of the problem in general. There are no active plans to support this, but mostly because it really hasn't been much of a problem so far.

Let's take TypeScript as an example: They have been pretty good at not breaking existing APIs. It's mostly new config values that are introduced, while old ones are deprecated. A recent example is the `extends` options. Sometime the type changed from accepting a single string to a list of string. Projen today supports both and leaves it up to the user to make sure it works. If possible we would add some validation. And for higher-level abstractions, projen would need to make sure that a correct configuration is generated.

--- 

Now, that doesn't mean that we won't have that problem in future. My thinking mostly has been that we would introduce a new `TypeScriptConfigV6` component that can be used instead. There's a few other pieces that would need to fall in place for this to be working, but it's not unfathomable. There's also ideas to split the library up into smaller, more focused pieces. In that world we could then release a `@projen/typescriptV6` package (or a v6 for said package). No separate registry would be needed as it would all reuse the existing package infrastructure. 

---

PS: Hello to Scotland from Scotland! 🏴󠁧󠁢󠁳󠁣󠁴󠁿

**Reply by janeklb** [reply]
Awesome - thanks for engaging and the thorough reply!
Encouraging to hear that it hasn't been a problem so far.

Separate "components-by-version" or separate versioned libraries (housing those components) would do the trick. A full on registry likely overkill; I was envisaging a future where everyone uses projen for everything.. but that bridge can be crossed when needed :)

**Reply by mrgrain** [reply]
Ideally there would be a projen project to create your own registry then! 😍

**Reply by thoroc** [reply]
Can we not publish construct on https://constructs.dev/? Or am I misunderstanding the question here?

**Reply by mrgrain** [reply]
> Can we not publish construct on https://constructs.dev/? Or am I misunderstanding the question here?

Yes we probably could.
